container.go

```diff
@@ -62,24 +56,21 @@ func (c *Container) Invoke(t interface{}) error {
 	ctype := reflect.TypeOf(t)
 	switch ctype.Kind() {
 	case reflect.Func:
-		c.RLock()
-		args, err := c.getArguments(ctype)
-		c.RUnlock()
+		args, err := c.Graph.ConstructorArguments(ctype)
 		if err != nil {
 			return err
 		}
 		cv := reflect.ValueOf(t)
 
 		// execute the provided func
 		values := cv.Call(args)
+
 		if len(values) > 0 {
-			c.Lock()
-			defer c.Unlock()
 			err, _ := values[len(values)-1].Interface().(error)
 			for _, v := range values {
 				switch v.Type().Kind() {
 				case reflect.Ptr:
-					c.insertObjectToGraph(v, v.Type())
+					c.Graph.InsertObject(v, v.Type())
```

Comment:
looks like we should strip type from the InsertObject, container should get it form the object itself.